pyiceberg/io/pyarrow.py

```diff
@@ -638,10 +634,165 @@ def visit_or(self, left_result: pc.Expression, right_result: pc.Expression) -> p
+        self._handle_nan_unmentioned(term)
+
+    def visit_not_null(self, term: BoundTerm[Any]) -> None:
+        self._handle_explicit_is_null_or_not(term)
+        self._handle_nan_unmentioned(term)
+
+    def visit_equal(self, term: BoundTerm[Any], literal: Literal[Any]) -> None:
+        self._handle_null_unmentioned(term)
+        self._handle_nan_unmentioned(term)
+
+    def visit_not_equal(self, term: BoundTerm[Any], literal: Literal[Any]) -> None:
+        self._handle_null_unmentioned(term)
+        self._handle_nan_unmentioned(term)
+
+    def visit_greater_than_or_equal(self, term: BoundTerm[Any], literal: Literal[Any]) -> None:
+        self._handle_null_unmentioned(term)
+        self._handle_nan_unmentioned(term)
+
+    def visit_greater_than(self, term: BoundTerm[Any], literal: Literal[Any]) -> None:
+        self._handle_null_unmentioned(term)
+        self._handle_nan_unmentioned(term)
+
+    def visit_less_than(self, term: BoundTerm[Any], literal: Literal[Any]) -> None:
+        self._handle_null_unmentioned(term)
+        self._handle_nan_unmentioned(term)
+
+    def visit_less_than_or_equal(self, term: BoundTerm[Any], literal: Literal[Any]) -> None:
+        self._handle_null_unmentioned(term)
+        self._handle_nan_unmentioned(term)
+
+    def visit_starts_with(self, term: BoundTerm[Any], literal: Literal[Any]) -> None:
+        self._handle_null_unmentioned(term)
+        self._handle_nan_unmentioned(term)
+
+    def visit_not_starts_with(self, term: BoundTerm[Any], literal: Literal[Any]) -> None:
+        self._handle_null_unmentioned(term)
+        self._handle_nan_unmentioned(term)
+
+    def visit_true(self) -> None:
+        return
+
+    def visit_false(self) -> None:
+        return
+
+    def visit_not(self, child_result: pc.Expression) -> None:
+        return
+
+    def visit_and(self, left_result: pc.Expression, right_result: pc.Expression) -> None:
+        return
+
+    def visit_or(self, left_result: pc.Expression, right_result: pc.Expression) -> None:
+        return
+
+    def collect(
+        self,
+        expr: BooleanExpression,
+    ) -> None:
+        """Collect the bound references categorized by having at least one is_null or is_not_null in the expr and the remaining."""
+        boolean_expression_visit(expr, self)
+
+
 def expression_to_pyarrow(expr: BooleanExpression) -> pc.Expression:
     return boolean_expression_visit(expr, _ConvertToArrowExpression())
 
 
+def _expression_to_complementary_pyarrow(expr: BooleanExpression) -> pc.Expression:
+    """Complementary filter conversion function of expression_to_pyarrow.
+
+    Could not use expression_to_pyarrow(Not(expr)) to achieve this complementary effect because ~ in pyarrow.compute.Expression does not handle null.
+    """
+    collector = _NullNaNUnmentionedTermsCollector()
+    collector.collect(expr)
+
+    def _downcast_term_to_reference(bound_terms: Set[BoundTerm[Any]]) -> Set[BoundReference[Any]]:
```

Comment:
Hi I want to understand more about the difference between BoundReference and BoundTerm.

Because class _NullNaNUnmentionedTermsCollector is inheriting from BoundBooleanExpressionVisitor[Any], the visit functions have to deal with BoundTerm. And in order to get the field reference to build IsNull and IsNaN, we need to do this ugly downcasting.

I saw BoundReference is the only child class of BoundTerm so far. 

What should be the best practice here? Should I create new abstract class of visitor that works with reference? Or a simpler change could be that I move the typecheck and downcasting into _handle_explicit_is_null_or_not() and its 3 siblings to make _NullNaNUnmentionedTermsCollector to have set[BoundReference[Any]] as collected results rather than set[BoundTerm[Any]]?

Reply:
Hi Adrian, I don't have much knowledge regarding the difference between a Reference and a Term and why we organized the classes this way, but I took a stab at resolving this issue without requiring a new Visitor. I've created a PR to your branch - please have a look and let me know if this works for you!

https://github.com/jqin61/iceberg-python/pull/5

Reply:
ah i see,  BoundPredicate's interface is BoundTerm, so it is already upcast. I was creating non-existing trouble for myself. Thanks a lot for the catch!

Reply:
fixed the types in tests and removed downcast in [b8aa4d8](https://github.com/apache/iceberg-python/pull/955/commits/b8aa4d888aec8ef88941fd48651fc9835ce73248)